Port MIOpenUtilKernel3 from OpenCL to HIP

EwanC opened the pull request:
## Motivation

This PR is part of the effort to port remaining OpenCL kernels in MIOpen to HIP. In particular porting `MIOpenUtilKernel3.cl` for subsampling and upsampling from OpenCL-C to HIP, including the 3 solves that use it.

## Technical Details

* Updated kernel data types to use `FLOAT` definition from the `float_types.h`.
* Updated host code to use `KernelBuildParameters` for setting macro defines.
* Inline implementation of `iMul` using straightforward port from OpenCL-C version in `math_ops.h`
* Updated solver code to add subsample/upsample kernels through the `Handle` API rather than through `ConvSolution.construction_params`, allowing the solution to take advantage of caching and avoiding test failures in `test_db_sync` from missing KDB entries that are otherwise seen without this update.

## Test Plan

Tested using `MIOpenDriver conv` for each of the following solutions `ConvAsm1x1U`, `ConvAsmBwdWrW1x1`, `ConvOclBwdWrW1x1` with `fp32`, `fp16`, and `bfp16` in different variations of shapes that allow the solutions to be applicable.

An extra configuration was also added to the existing `test_unit_conv_solver_ConvAsm1x1U` and `test_unit_conv_solver_ConvOclBwdWrW1x1` smoke tests with a shape that stresses the subsample/upsample kernels from these solutions.

## Test Result

Tests pass on gfx906, gfx908, and gfx90a.

## Submission Checklist

- [x] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

zsotakal commented on projects/miopen/src/solver/conv/conv_ocl_dir2D_bwdWrW_1x1.cpp:
```diff
@@ -461,10 +474,9 @@ ConvSolution ConvOclBwdWrW1x1::GetSolution(const ExecutionContext& ctx,
         if(n_passes == 2)
         {
-            result.invoker_factory = [ws_sz](const std::vector<Kernel>& kernels) {
+            result.invoker_factory = [ws_sz, ss_kernel_info](const std::vector<Kernel>& kernels) {
```
nitpick: would it be possible to use move semantics on ss_kernel_info? As I see we do a copy here which could be avioded.

EwanC replied:
Good observation, I've updated the code to use `ss_kernel_info = std::move(ss_kernel_info)` in the lamdba captures for kernel info

BalintCsala commented:
LGTM

muralinr commented on projects/miopen/src/kernels/MIOpenUtilKernels3.cpp:
```diff
@@ -0,0 +1,74 @@
+                                                             FLOAT* __restrict out)
+{
+    const unsigned int stack_pos = blockIdx.x * LOCAL_SIZE_X + threadIdx.x;
+    const unsigned int batch_id  = blockIdx.y * LOCAL_SIZE_Y + threadIdx.y;
```
Bounds Checking - CRITICAL**
// No validation that stack_pos or batch_id are within valid ranges
const unsigned int stack_pos = blockIdx.x * LOCAL_SIZE_X + threadIdx.x;
const unsigned int batch_id  = blockIdx.y * LOCAL_SIZE_Y + threadIdx.y;

If grid dimensions are miscalculated, threads may access out-of-bounds memory.

EwanC replied:
There was no validation in the existing OpenCL-C kernel, so i'm reluctant to add it to add extra code to the HIP code when the goal for now is to make them functionally identical.

muralinr replied:
ok.

muralinr commented on projects/miopen/src/kernels/MIOpenUtilKernels3.cpp:
```diff
@@ -0,0 +1,74 @@
+    unsigned int out_y   = pix_pos / out_stride_aligned;
+    unsigned int out_x   = iMod(pix_pos, out_y, out_stride_aligned) * WRITE_UNIT;
+
+    unsigned int out_off = batch_id * IN_BATCH_STRIDE + stack_pos * WRITE_UNIT;
```
**Integer Overflow Risk**
unsigned int out_off = batch_id * IN_BATCH_STRIDE + stack_pos * WRITE_UNIT;

Large tensor dimensions could cause 32-bit unsigned overflow. Should consider using `size_t` or `unsigned long long`.

EwanC replied:
In the solvers there are generally checks in `IsApplicable()`  that `n_c_h_w < std::pow(2, 29)`, so i imagine that would be upper limit on tensor size.

Would document a limitation like that in the kernel code be satisfactory, or would you like `uint64_t` offsets to be safe?

EwanC replied:
Good catch, I've added a code comment highlighting this assumption, but as this was an issue in the OpenCL kernel I think it's out of scope for addressing in this PR.

muralinr replied:
ok.

muralinr commented on projects/miopen/src/kernels/MIOpenUtilKernels3.cpp:
```diff
@@ -0,0 +1,74 @@
+
+#include "float_types.h"
+
+inline __device__ unsigned int iMod(unsigned int v, unsigned int u, unsigned int d)
```
iMod Function Correctness:

inline __device__ unsigned int iMod(unsigned int v, unsigned int u, unsigned int d)
{
    return v - __mul24(u, d);  // Assumes u = v / d
}

The function name suggests modulo, but it's actually computing `v - u*d`. The caller must ensure `u = v/d` for correctness. This might be error-prone.

muralinr replied:
__mul24 Limitation:
return v - __mul24(u, d);

`__mul24` only works correctly for 24-bit integers. If `u` or `d` exceed 2^24, results are undefined.

EwanC replied:
I agree the prior divide should really be part of the function if it's an expectation for the function to work correctly. I've moved the divide inside the function such that it can return the quotient as an output parameter and renamed it to `iRemquo`. 

I've not addressed the `__mul24` point because that was the designed of the original CL kernel, but I have added a comment to make this assumption explicit in the code.

muralinr replied:
ok.

muralinr commented on projects/miopen/src/kernels/MIOpenUtilKernels3.cpp:
```diff
@@ -0,0 +1,74 @@
+    return v - __mul24(u, d);
+}
+
+constexpr unsigned int out_channel_stride_aligned = OUT_CHANNEL_STRIDE / WRITE_UNIT;
```
Alignment Assumptions:
constexpr unsigned int out_channel_stride_aligned = OUT_CHANNEL_STRIDE / WRITE_UNIT;

No compile-time assertion that `OUT_CHANNEL_STRIDE` is divisible by `WRITE_UNIT`. Misalignment causes incorrect indexing.

EwanC replied:
Agree we can assert this at compile time, I've added a static assert before each `constexpr` divide.

muralinr replied:
good. thank you.

muralinr commented on projects/miopen/src/kernels/MIOpenUtilKernels3.cpp:
```diff
@@ -0,0 +1,74 @@
+constexpr unsigned int in_stride_aligned         = IN0_STRIDE / WRITE_UNIT;
+
+extern "C" __global__
+__launch_bounds__(LOCAL_SIZE_X* LOCAL_SIZE_Y) void UpSample(const FLOAT* __restrict in,
```
Uninitialized Output Memory:
The UpSample kernel only writes to strided positions. If the output buffer isn't pre-zeroed, it will contain garbage values in the gaps.

EwanC replied:
Well spotted, the solver is indeed pre-zeroing the memory before calling this kernel. I've added a comment pointing that output and referenced the todo that exists there for optimizing this further.

muralinr replied:
ok.

muralinr commented:
Here are overall comments:

1. **Add bounds checking** at kernel entry
2. **Use 64-bit arithmetic** for offset calculations
3. **Add static_assert** for alignment requirements
4. **Document iMod** behavior or replace with standard modulo
5. **Consider memory coalescing** - UpSample may benefit from transposed access patterns
6. **Add kernel parameter validation** in host code
7. **Pre-zero output buffer** for UpSample or make it explicit in documentation

muralinr commented:
Memory Access Pattern:
for(unsigned int i = 0; i < WRITE_UNIT; ++i, in_ptr += FILTER0_STRIDE0, out_ptr++)
**Issue**: 
- **SubSample**: Strided reads, coalesced writes (good for downsampling)
- **UpSample**: Coalesced reads, strided writes (may cause poor memory performance)

BradPepersAMD commented:
@EwanC the changes here look to be functionally identical to what the OCL code was doing with some improvements to kernel argument handling.  @muralinr mentions some good change to fix issues that I believe were all issues in the original code as well.  Since the goal was OCL to HIP functionally identical, I'm ok with not fixing all the issues found by Murali and instead make that a separate PR to improve SoftMax though if you find any of them are trivial to do now, it would be good to have those improvements!

EwanC commented:
> @EwanC the changes here look to be functionally identical to what the OCL code was doing with some improvements to kernel argument handling. @muralinr mentions some good change to fix issues that I believe were all issues in the original code as well. Since the goal was OCL to HIP functionally identical, I'm ok with not fixing all the issues found by Murali and instead make that a separate PR to improve SoftMax though if you find any of them are trivial to do now, it would be good to have those improvements!

Thanks for the input Brad, I've addressed the stylistic issues that should affect the compiled code and added code comments highlighting the rest. @muralinr Could you please re-review with respect to the scope of the PR being functional equivalence.

Edit: Found this CI fail from the static assert not passing for the `Upsample` kernel when the solver was only intending on running `SubSample` from the compiled program. Fixed by splitting `MIOpenUtilKenel3` into two separate files for each kernel so that we can keep the static_asserts.
```
MIOpen(HIP): Warning [BuildHip] MIOpenUtilKernels3.cpp:58:15: error: static assertion failed due to requirement '(25 % 3) == 0'
```

EwanC commented:
Hey @muralinr, this PR is blocked on your review as you've requested changes. I don't think there are any open comment threads so could you continue discussions or approve the PR.

muralinr reviewed (commented):
LGTM. I am fine with doing my review comments as part of new PRs as current PR is limiting scope from OCL to HIP conversions.
